### DISCUSSION TITLE
Need example on GetConnect usage

### DISCUSSION THREAD

**Discussion by Darksoulsong** [jonataslaw/getx#938] [Q&A]
I just discovered GetConnect class in the docs and I'm trying to use it, but I can't figure how to get that properly imported. VSCode simply doesn't know which GetX file should be imported to get GetConnect working. Can you guys give me directions?

I tried many different imports, but none of them seems to have the GetConnect class definition.

import 'package:get/utils.dart';
import 'package:get/get.dart';
import 'package:get/get_core/get_core.dart';
import 'package:get/get_state_manager/get_state_manager.dart';
import 'package:get/get_rx/get_rx.dart';
import 'package:get/get_instance/get_instance.dart';

**Comment by jonataslaw** [comment]
import 'package:get/get.dart';

**Reply by Darksoulsong** [reply]
Thank you @jonataslaw ! But I have tried that already. VSCode keeps giving me a warning about a unused import and "Classes can only extend other classes." on GetConnect.

Am I missing something?

![image](https://user-images.githubusercontent.com/2078582/102931030-5e14ff80-447c-11eb-9947-b1e0f014b03a.png)

**Comment by jonataslaw** [comment]
Update `get` package on your pubspec to last version

**Reply by Darksoulsong** [reply]
This did the trick. I had to bump ^3.16.0 to ^3.24.0. 
Thank you!

**Comment by Darksoulsong** [comment]
I'm not sure about how GetConnect should be used. Do I need to register it somewhere to be able to use it? I'm importing it directly on my controller like so:

// orders controller
```
var provider = new OrdersProvider();
var result = await provider.getOrders(type);
```

But the providers "onInit" method is never called on instantiation and I'm getting "ArgumentError (Invalid argument(s): No host specified in URI" exception.

Here goes the related code:
```
class Order {
  String id;
  String total;

  Order(
    {this.id,
    this.total,
    });

  static fromJson(json) {
    return json != null
        ? Order(
            id: json['id'],
            total: json['total'],
          )
        : null;
  }
}

class OrdersProvider extends GetConnect {
  Future<Response> getOrders(String range) {
    Map body = {
      'someParam': 'foo',
    };

    return post('/orders', body);
  }

  @override
  void onInit() {
    httpClient.baseUrl = API_URL;
    httpClient.defaultDecoder = Order.fromJson;
  }
}

class OrdersControllerG extends GetxController
    with SingleGetTickerProviderMixin {
  var appBarTitle = 'Some title';

  RxBool loading = false.obs;

  RxList<dynamic> ordersData = [].obs;  

  @override
  void onInit() {
    getData();    
    super.onInit();
  }

  getData() async {
    loading.toggle();
    ordersData.assignAll([]);
    
    var provider = new OrdersProvider();
    var result = await provider.getDeclinedOrders(type);

    print(result.body);
    loading.toggle();
    ordersData.assignAll(result.body['data']);
  }
}
```

It would be awesome if we had a complete example of how to use it properly.

**Reply by eduardoflorence** [reply]
Remove `var provider = new OrdersProvider();` 
and before `var appBarTitle..."` put `final provider = Get.put(OrdersProvider());`

If you use Bindings, do Get.put on it and inject in the controller constructor, as this is a good practice

**Reply by smjxpro** [reply]
Isn't putting Get.put() everywhere the same thing as instantiating dependencies everywhere the same thing?

I think that defeats the purpose of dependency injection.

**Reply by starlimit** [reply]
Using Bindings helps to avoid `Get.put(MyProvider());` everywhere. You can have it in a single location and reference it by including the bindings. see a sample usage [here](https://github.com/starlimit/Flutter-GetX)

**Comment by yang-lile** [comment]
How to use `GetHttpClient.addAuthenticator`, and why don't have `GetHttpClient.removeAuthenticator`? We need code comment and usage document! @jonataslaw

**Reply by jasonlaw** [reply]
I use addAuthenticator to refresh my bearer token when it is expired and return with 401 status code.
Here is my code example:
~~~dart
App.connect.addAuthenticator<void>((request) async {
    final token = CredentialIdentity.refreshToken;
    if (token != null) {
      await '/api/login/refresh'
          .api()
          .post(headers: {'x-refresh-token': token});
    }
    return request;
  });
~~~
